<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

- In Deep learning, to avoid model over-fitting, we uses Dropout Regularization method.
- Randomly dropping few neurons/Hidden units from overfit models hidden layer, this method is called Dropout Regularization.

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

- Here Output has category as metal and rocks

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

https://www.youtube.com/watch?v=lcI8ukTUEbo&list=PLeo1K3hjS3uu7CxAacxVndI4bE_o3BDtO&index=20

In [31]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [32]:
import warnings
warnings.filterwarnings('ignore')

In [33]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

,0,1,2,3,4,5,6,7,8,9,...,51,52,53,54,55,56,57,58,59,60
22,0.0099,0.0484,0.0299,0.0297,0.0652,0.1077,0.2363,0.2385,0.0075,0.1882,...,0.0173,0.0149,0.0115,0.0202,0.0139,0.0029,0.0160,0.0106,0.0134,R
81,0.0100,0.0194,0.0155,0.0489,0.0839,0.1009,0.1627,0.2071,0.2696,0.2990,...,0.0130,0.0073,0.0077,0.0075,0.0060,0.0080,0.0019,0.0053,0.0019,R
144,0.0299,0.0688,0.0992,0.1021,0.0800,0.0629,0.0130,0.0813,0.1761,0.0998,...,0.0459,0.0277,0.0172,0.0087,0.0046,0.0203,0.0130,0.0115,0.0015,M
200,0.0131,0.0387,0.0329,0.0078,0.0721,0.1341,0.1626,0.1902,0.2610,0.3193,...,0.0150,0.0076,0.0032,0.0037,0.0071,0.0040,0.0009,0.0015,0.0085,M
126,0.0715,0.0849,0.0587,0.0218,0.0862,0.1801,0.1916,0.1896,0.2960,0.4186,...,0.0153,0.0121,0.0096,0.0196,0.0042,0.0066,0.0099,0.0083,0.0124,M


In [34]:
df.shape

(208, 61)

In [35]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [36]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [37]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

> Here 111 are metal and 97 are Rocks actegory and equally distributed

In [38]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [39]:
#Convert category to numeric by onehot encoding
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

,R
23,1
192,0
111,0
141,0
106,0


In [40]:
type(y)

pandas.core.frame.DataFrame

In [41]:
y['R'].value_counts()

0    111
1     97
Name: R, dtype: int64

In [42]:
X.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0232,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0125,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0033,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0241,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0156,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094


## Train test split

In [43]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [44]:
X_train.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
67,0.0368,0.0403,0.0317,0.0293,0.0820,0.1342,0.1161,0.0663,0.0155,0.0506,...,0.0058,0.0091,0.0160,0.0160,0.0081,0.0070,0.0135,0.0067,0.0078,0.0068
14,0.0124,0.0433,0.0604,0.0449,0.0597,0.0355,0.0531,0.0343,0.1052,0.2120,...,0.0078,0.0083,0.0057,0.0174,0.0188,0.0054,0.0114,0.0196,0.0147,0.0062
164,0.0163,0.0198,0.0202,0.0386,0.0752,0.1444,0.1487,0.1484,0.2442,0.2822,...,0.0027,0.0077,0.0026,0.0031,0.0083,0.0020,0.0084,0.0108,0.0083,0.0033
179,0.0394,0.0420,0.0446,0.0551,0.0597,0.1416,0.0956,0.0802,0.1618,0.2558,...,0.0118,0.0146,0.0040,0.0114,0.0032,0.0062,0.0101,0.0068,0.0053,0.0087
19,0.0126,0.0149,0.0641,0.1732,0.2565,0.2559,0.2947,0.4110,0.4983,0.5920,...,0.0153,0.0092,0.0035,0.0098,0.0121,0.0006,0.0181,0.0094,0.0116,0.0063


# Using Deep Learning Model

## Model without Dropout Layer

In [45]:
import tensorflow as tf
from tensorflow import keras

In [46]:
# Here 60,30,15 ouput dim are randomly selected,last 1 is the actual based on output req. 
#Here op is binary result, so chose 1 hidden unit as o/p

model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# as its binary classification problem we have used 'binary_crossentropy'

model.fit(X_train, y_train, epochs=100, batch_size=8)  # Here used Mini batch Gradient Descent with batch size=8

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6843 - accuracy: 0.5385
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6595 - accuracy: 0.6474
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6398 - accuracy: 0.6859
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6109 - accuracy: 0.7756
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5776 - accuracy: 0.7821
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5431 - accuracy: 0.7949
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5087 - accuracy: 0.7821
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4740 - accuracy: 0.7692
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4624 - accuracy: 0.8077
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4342 - accuracy: 0.8333
Epoch 11/

20/20 [==============================] - 0s 3ms/step - loss: 0.0081 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0079 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0107 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0095 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0103 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0105 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0064 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0052 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0051 - accuracy: 1.0000
Epoch 92/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0055 - accuracy: 1.0000
Epoch 93/100


In [47]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 2ms/step - loss: 1.0624 - accuracy: 0.7692


[1.0624363422393799, 0.7692307829856873]

> Here accuracy of train data  =1 , but with test data its 0.7692 --> So this looks like overfit model

Training Accuracy >>> Test Accuracy

In [48]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[1.7055223e-06 9.8673832e-01 9.9651957e-01 8.0421785e-05 1.0000000e+00
 9.9961841e-01 3.7274331e-02 1.0000000e+00 4.3368340e-04 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [49]:
y_test[:10]

,R
186,0
155,0
165,0
200,0
58,1
34,1
151,0
18,1
202,0
62,1


In [50]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      0.85      0.79        27
           1       0.81      0.68      0.74        25

    accuracy                           0.77        52
   macro avg       0.78      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



### Model with Dropout Layer
- Here dropout applied after each layer, here 0.5 means 50% hiddenunits are dropped randomly

In [51]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5), 
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7823 - accuracy: 0.4551
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7161 - accuracy: 0.5513
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7046 - accuracy: 0.4872
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7161 - accuracy: 0.4808
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6998 - accuracy: 0.4744
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6910 - accuracy: 0.5705
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6965 - accuracy: 0.5000
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6827 - accuracy: 0.5962
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6975 - accuracy: 0.5128
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6836 - accuracy: 0.5192
Epoch 11/

20/20 [==============================] - 0s 3ms/step - loss: 0.4180 - accuracy: 0.8077
Epoch 84/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3279 - accuracy: 0.8974
Epoch 85/100
20/20 [==============================] - 0s 4ms/step - loss: 0.3618 - accuracy: 0.8333
Epoch 86/100
20/20 [==============================] - 0s 4ms/step - loss: 0.4141 - accuracy: 0.8526
Epoch 87/100
20/20 [==============================] - 0s 4ms/step - loss: 0.3950 - accuracy: 0.8462
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3936 - accuracy: 0.8462
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3676 - accuracy: 0.8462
Epoch 90/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3935 - accuracy: 0.8333
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3330 - accuracy: 0.8526
Epoch 92/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3429 - accuracy: 0.8718
Epoch 93/100


In [52]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 4ms/step - loss: 0.4615 - accuracy: 0.7692


[0.4615395665168762, 0.7692307829856873]

> Here accuracy of train data  = 0.8269 , and with test data its 0.7692 --> So this looks like NOT overfit model, both accuracy are almost nearer

- So this dropout helped to reduce overfit problem. This mainly used in computer vision, where complex neural network with high overfit problem


Training Accuracy is still good but Test Accuracy Improved

In [53]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[4.0942430e-04 8.5563195e-01 9.3275410e-01 3.1206548e-02 9.9641812e-01
 8.9563310e-01 2.8651327e-01 9.9556708e-01 5.7962388e-02 9.9812460e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [54]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.76      0.81      0.79        27
           1       0.78      0.72      0.75        25

    accuracy                           0.77        52
   macro avg       0.77      0.77      0.77        52
weighted avg       0.77      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.75 to 0.77**
- Here dropout applied randomly only random hidden units, so accuracy will be different in each run